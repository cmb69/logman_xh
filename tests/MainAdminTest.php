<?php

namespace Logman;

use ApprovalTests\Approvals;
use Logman\Infra\View;
use Logman\Model\Entry;
use Logman\Model\Logfile;
use PHPUnit\Framework\MockObject;
use PHPUnit\Framework\TestCase;

class MainAdminTest extends TestCase
{
    public function testDisplaysLogfile(): void
    {
        $logfile = $this->createMock(Logfile::class);
        $logfile->expects($this->once())->method("find")->willReturn([$this->loginSuccessEntry()]);
        $sut = $this->sut($this->conf(), $logfile, $this->view());
        $_GET = [
            "action" => "plugin_text",
            "logman_timestamp" => "2025",
            "logman_level" => "info",
            "logman_module" => "XH",
            "logman_category" => "login",
            "logman_description" => "from"
        ];
        Approvals::verifyHtml($sut());
    }

    /** <https://github.com/cmb69/logman_xh/issues/1> */
    public function testZeroMaxEntriesDisplaysEntries(): void
    {
        $conf = $this->conf();
        $conf["entries_max"] = "0";
        $logfile = $this->createMock(Logfile::class);
        $logfile->expects($this->once())->method("find")->with($this->anything(), PHP_INT_MAX)
            ->willReturn([$this->loginSuccessEntry()]);
        $sut = $this->sut($conf, $logfile, $this->view());
        $_GET = [
            "action" => "plugin_text",
        ];
        $this->assertStringContainsString("2023-01-30 14:00:05", $sut());
    }

    public function testDisplaysDeleteConfirmation(): void
    {
        $logfile = $this->createMock(Logfile::class);
        $sut = $this->sut($this->conf(), $logfile, $this->view());
        $_GET = [
            "action" => "delete",
            "logman_count" => "1",
            "logman_timestamp" => "2025",
            "logman_level" => "info",
            "logman_module" => "XH",
            "logman_category" => "login",
            "logman_description" => "from"
        ];
        Approvals::verifyHtml($sut());
    }

    public function testDeletesEntries(): void
    {
        $logfile = $this->createMock(Logfile::class);
        $logfile->expects($this->once())->method("delete")->willReturn(1);
        $sut = $this->sut($this->conf(), $logfile, $this->view());
        $sut->expects($this->once())->method("redirect");
        $_SERVER["QUERY_STRING"] = "";
        $_GET = [
            "action" => "delete",
            "logman_count" => "1",
            "logman_timestamp" => "2025",
            "logman_level" => "info",
            "logman_module" => "XH",
            "logman_category" => "login",
            "logman_description" => "from"
        ];
        $_POST = [
            "logman_do" => "",
        ];
        $sut();
    }

    public function testDisplaysNumberOfDeletedEntries(): void
    {
        $logfile = $this->createMock(Logfile::class);
        $logfile->expects($this->once())->method("find")->willReturn([$this->loginSuccessEntry()]);
        $sut = $this->sut($this->conf(), $logfile, $this->view());
        $_SERVER["QUERY_STRING"] = "";
        $_GET = [
            "action" => "",
            "logman_count" => "1",
            "logman_deleted" => "17",
            "logman_timestamp" => "2025",
            "logman_level" => "info",
            "logman_module" => "XH",
            "logman_category" => "login",
            "logman_description" => "from"
        ];
        $this->assertStringContainsString("17 entries deleted!", $sut());
    }

    private function sut(array $conf, Logfile $logfile, View $view)
    {
        return $this->getMockBuilder(MainAdmin::class)
            ->setConstructorArgs([$conf, $logfile, $view])
            ->onlyMethods(["redirect"])
            ->getMock();
    }

    private function conf(): array
    {
        return XH_includeVar("./config/config.php", "plugin_cf")["logman"];
    }

    private function view(): View
    {
        return new View("./views/", XH_includeVar("./languages/en.php", "plugin_tx")["logman"]);
    }

    private function loginSuccessEntry(): Entry
    {
        return new Entry("2023-01-30 14:00:05", "info", "XH", "login", "login from ::1");
    }
}
