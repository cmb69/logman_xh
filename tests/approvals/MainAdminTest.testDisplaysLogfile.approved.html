
<h1>Logman â€“ Log file</h1>
<form class="logman_form" method="get">
  <input type="hidden" name="selected" value="logman">
  <input type="hidden" name="admin" value="plugin_main">
  <input type="hidden" name="action" value="plugin_text">
  <div class="logman_table_wrapper">
    <table>
      <thead>
        <tr>
          <th>Timestamp</th>
          <th>Level</th>
          <th>Module</th>
          <th>Category</th>
          <th>Description</th>
        </tr>
        <tr>
          <td><input type="search" name="logman_timestamp" value="2025" placeholder="until" list="logman_months"></td>
          <td><input type="search" name="logman_level" value="info" placeholder="is" list="logman_levels"></td>
          <td><input type="search" name="logman_module" value="XH" placeholder="is" list="logman_modules"></td>
          <td><input type="search" name="logman_category" value="login" placeholder="is" list="logman_categories"></td>
          <td><input type="search" name="logman_description" value="from" placeholder="contains"></td>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>2023-01-30 14:00:05</td>
          <td>info</td>
          <td>XH</td>
          <td>login</td>
          <td>login from ::1</td>
        </tr>
      </tbody>
    </table>
  </div>
  <p class="logman_buttons"><button>Filter</button></p>
  <datalist id="logman_months">
    <option value="2023-01"></option>
  </datalist>
  <datalist id="logman_levels">
    <option value="info"></option>
  </datalist>
  <datalist id="logman_modules">
    <option value="XH"></option>
  </datalist>
  <datalist id="logman_categories">
    <option value="login"></option>
  </datalist>
</form>
